---
title: 'Aurora Tool Report'
output:
    html_document:
      highlight: pygments
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(error = TRUE, echo = FALSE)
```

```{css echo=FALSE}
# code chunks scrollable
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```


```{r, echo=FALSE}
# to make the css theme to work, <link></link> tags cannot be added directly 
# as <script></script> tags as below.
# it has to be added using a code chunk with the htmltool functions!!!
css_link = tags$link()
css_link$attribs = list(rel="stylesheet", href="vakata-jstree-3.3.5/dist/themes/default/style.min.css")
css_link
```

```{r, eval=FALSE, echo=FALSE}
# this code chunk is purely for adding comments
# below is to add jQuery and jstree javascripts
```
<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="vakata-jstree-3.3.5/dist/jstree.min.js"></script>

---
# javascript code below is to build the file tree interface
# see this for how to implement opening hyperlink: https://stackoverflow.com/questions/18611317/how-to-get-i-get-leaf-nodes-in-jstree-to-open-their-hyperlink-when-clicked-when
---
<script>
  $(function () {
    // create an instance when the DOM is ready
    $('#jstree').jstree().bind("select_node.jstree", function (e, data) {
     window.open( data.node.a_attr.href, data.node.a_attr.target )
    });
  });
</script>

---
# ADD YOUR DATA ANALYSIS CODE AND MARKUP TEXT BELOW TO EXTEND THIS R MARKDOWN FILE
---



## Load Data

```{r}
VCF <- read.vcfR(opt$X_v) # import the VCF in R
```

## Create a Genotype Matrix

The VCF object contains several additional information concerning genetic variant quality that are not needed in this exercise. We want to obtain a Genotype Matrix out of the VCF object, we can do it as shown here below:  

```{r}
GT <- vcfR2loci(VCF)

GT[1:10,1:10] # The GT matrix shows the genotype for each individual (row) and genetic variant (column)

dim(GT) 
```

For our calculations, we prefer having the genotypes coded as 0,1,2 (homozygous 0/0, heterozygous, homozygous 1/1). The code here below performs this transformation creating the SNPs matrix. 

```{r}
GT <- as.matrix(GT)

GT[GT=='0/0'] <- 0
GT[GT=='0/1'] <- 1
GT[GT=='1/0'] <- 1
GT[GT=='1/1'] <- 2

SNPS <- GT
SNPS <- apply(SNPS,2,as.numeric)
rownames(SNPS) <- rownames(GT)
```


The SNPS matrix has the genotype values coded as 0,1,2:

```{r}
SNPS[1:10,1:10]
```


## Check for genetic diversity metrics

Before starting, we create a fSNPS matrix identical to the SNPS matrix. We will perform all the calculation on the fSNPS matrix keeping the SNPS matrix as backup.

```{r}
fSNPS <- SNPS
```


There are R packages that allow to compute this metrics directly with built-in function. In this exercise we use the hand-written formula to better show the meaning of each procedure.

### Missingness by SNP

For each SNP, it computes the % of missing observation across the individuals. 

```{r}
MNsnp <- apply(fSNPS, 2, function(x) { sum(is.na(x))/length(x) })
hist(MNsnp, breaks=10) # We can observe that most of the SNPS have a Missingness below 0.01 (i.e. ~ 2 individual ot of 160).
fSNPS <- fSNPS[,MNsnp<0.01] # We apply a 1% threshold
write.csv(fSNPS, file = "fSNPS.csv", quote = FALSE, col.names = TRUE)
```
