---
title: 'Aurora Tool Report'
output:
    html_document:
      highlight: pygments
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(error = TRUE, echo = FALSE)
```

```{css echo=FALSE}
# code chunks scrollable
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```


```{r, echo=FALSE}
# to make the css theme to work, <link></link> tags cannot be added directly 
# as <script></script> tags as below.
# it has to be added using a code chunk with the htmltool functions!!!
css_link = tags$link()
css_link$attribs = list(rel="stylesheet", href="vakata-jstree-3.3.5/dist/themes/default/style.min.css")
css_link
```

```{r, eval=FALSE, echo=FALSE}
# this code chunk is purely for adding comments
# below is to add jQuery and jstree javascripts
```
<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="vakata-jstree-3.3.5/dist/jstree.min.js"></script>

---
# javascript code below is to build the file tree interface
# see this for how to implement opening hyperlink: https://stackoverflow.com/questions/18611317/how-to-get-i-get-leaf-nodes-in-jstree-to-open-their-hyperlink-when-clicked-when
---
<script>
  $(function () {
    // create an instance when the DOM is ready
    $('#jstree').jstree().bind("select_node.jstree", function (e, data) {
     window.open( data.node.a_attr.href, data.node.a_attr.target )
    });
  });
</script>

---
# ADD YOUR DATA ANALYSIS CODE AND MARKUP TEXT BELOW TO EXTEND THIS R MARKDOWN FILE
---



# BDSS configuration file

First, we create a bdss configuration file `bdss.cfg` in the current directory.

```{r}
system('echo "[metadata_repository]" > bdss.cfg')
system('echo url=http://bdss.bioinfo.wsu.edu/ >> bdss.cfg')
```

# Download and extract reads

```{r 'download and extract reads'}
# create two directories, one for single end and the other for paired end SRA reads.
dir.create('se_read_files_directory')
dir.create('pe_read_files_directory')
# download and extract reads (single end)
sra_ids_se = strsplit(gsub(',', ' ', opt$X_s), ' ')[[1]]
sra_ids_se = sra_ids_se[sra_ids_se != '']
# loop through SRA accessions to download and extract reads.
for(id in sra_ids_se) {
    # build URL from SRA id
    url = paste0('ftp://ftp.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/',
                 substr(id, 1, 3), '/',
                 substr(id, 1, 6), '/', id, '/', id, '.sra')
    # download sra file with bdss
    bdss_command = paste0('/main/sites/galaxy/test-galaxy/tools/_conda/bin/bdss transfer -u ', url)
    system(bdss_command, intern = TRUE)
    # convert .sra to .fastq/.fasta
    if('FORMAT' == 'fasta') {
      command = paste0('fastq-dump --fasta -O se_read_files_directory ', id, '.sra')
    } else {
      command = paste0('fastq-dump -O se_read_files_directory ', id, '.sra')
    }
    cat('----convert SRA to fastq/fasta------\n')
    print(system(command, intern = TRUE))
}

# download and extract reads (paired end)
sra_ids_pe = strsplit(gsub(',', ' ', opt$X_p), ' ')[[1]]
sra_ids_pe = sra_ids_pe[sra_ids_pe != '']
# loop through SRA accessions to download and extract reads.
for(id in sra_ids_pe) {
    # build URL from SRA id
    url = paste0('ftp://ftp.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/',
                 substr(id, 1, 3), '/',
                 substr(id, 1, 6), '/', id, '/', id, '.sra')
    # download sra file with bdss
    bdss_command = paste0('/main/sites/galaxy/test-galaxy/tools/_conda/bin/bdss transfer -u ', url)
    system(bdss_command, intern = TRUE)
    # convert .sra to .fastq/.fasta
    if(opt$X_f == 'fasta') {
      command = paste0('fastq-dump --fasta --split-files -O pe_read_files_directory ', id, '.sra')
    } else {
      command = paste0('fastq-dump --split-files -O pe_read_files_directory ', id, '.sra')
    }
    cat('----convert SRA to fastq/fasta------\n')
    command_stdout = system(command, intern = TRUE)
    print(command_stdout)
    if(!(paste0(id, '_2.FORMAT') %in% list.files('pe_read_files_directory'))) {
      # this is not a paired end SRA file. The corresponding file will be deleted.
      cat(paste0(id, ' is not paired end SRA, the corresponding fastq/fasta file will deleted.'))
      system(paste0('rm pe_read_files_directory/', id, '_1.*'), intern = TRUE)
    }
    
}

cat('-----single end files----\n')
list.files('./se_read_files_directory')
cat('-----paired end files----\n')
list.files('./pe_read_files_directory')

cat('-----Renaming files------\n')
# rename files for paired end reads
old_files = paste0('./pe_read_files_directory/', list.files('./pe_read_files_directory'))
print(old_files)
new_files = gsub('_1', '_forward', old_files)
new_files = gsub('_2', '_reverse', new_files)
print(new_files)
file.rename(old_files, new_files)
```