---
title: 'HMMER search against Pfam database report'
output:
    html_document:
      highlight: pygments
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(error = TRUE, echo = TRUE)
```

```{css echo=FALSE}
# code chunks scrollable
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```


```{r, echo=FALSE}
# to make the css theme to work, <link></link> tags cannot be added directly 
# as <script></script> tags as below.
# it has to be added using a code chunk with the htmltool functions!!!
css_link = tags$link()
css_link$attribs = list(rel="stylesheet", href="vakata-jstree-3.3.5/dist/themes/default/style.min.css")
css_link
```

```{r, eval=FALSE, echo=FALSE}
# this code chunk is purely for adding comments
# below is to add jQuery and jstree javascripts
```
<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="vakata-jstree-3.3.5/dist/jstree.min.js"></script>

---
# javascript code below is to build the file tree interface
# see this for how to implement opening hyperlink: https://stackoverflow.com/questions/18611317/how-to-get-i-get-leaf-nodes-in-jstree-to-open-their-hyperlink-when-clicked-when
---
<script>
  $(function () {
    // create an instance when the DOM is ready
    $('#jstree').jstree().bind("select_node.jstree", function (e, data) {
     window.open( data.node.a_attr.href, data.node.a_attr.target )
    });
  });
</script>

---
# ADD YOUR DATA ANALYSIS CODE AND MARKUP TEXT BELOW TO EXTEND THIS R MARKDOWN FILE
---

# Import data

```{r}
library(stringr)
# get protein sequences
protein_sequences = readLines(opt$X_u)
# process sequence names
protein_sequences = str_replace_all(protein_sequences, '(^>.*?)_m.*', '\\1')
# write processed sequences into a file
write(protein_sequences, file = paste0(opt$X_d, '/protein_sequences.fasta'))

# get significant genes
sig_genes = read.table(opt$X_g, header = TRUE, stringsAsFactors = FALSE)[,1,drop=FALSE]
colnames(sig_genes) = 'Significant Genes'
# write significant genes to a file
write.table(sig_genes, file = paste0(opt$X_d, '/sig_genes.txt'), quote = FALSE, col.names = FALSE, row.names = FALSE)
# show the first few genes
knitr::kable(head(sig_genes))
```

# HMMER search against Pfam database

Extract protein sequences for significant genes

```{bash}
# extract protein sequences for significant genes
seqkit grep --pattern-file ${X_d}/sig_genes.txt ${X_d}/protein_sequences.fasta > ${X_d}/selected_protein_seqs.fasta
```

```{bash echo=FALSE}
sh ${TOOL_INSTALL_DIR}/build-and-run-job-scripts.sh
```

Run `hmmscan`

```{r echo=FALSE,warning=FALSE,results='asis'}
# display content of the job-script.sh file.
cat('```bash\n')
cat(readLines(paste0(Sys.getenv('REPORT_FILES_PATH'), '/hmmscan.sh')), sep = '\n')
cat('\n```')
```

Save hmmscan results into a file and display the first few results.

```{r echo=FALSE}
library(readr)
library(dplyr)
header = 'target_name,accession,query_name,e_value'
tblout = read_delim(paste0(opt$X_d, '/tblout.txt'), comment = '#', delim = '\t', col_names = FALSE)
tblout = c(header, str_replace_all(tblout$X1, '(.*?)\\s+(.*?)\\s+(.*?)\\s+(.*?)\\s+(.*?)\\s+.*', '\\1,\\2,\\3,\\5'))
if (opt$X_b) {
  tblout = group_by(tblout, target_name) %>%
    slice(which.min(e_value))
}
write.table(tblout, file = paste0(opt$X_d, '/tblout.csv'), col.names = FALSE, row.names = FALSE, quote = FALSE)
knitr::kable(head(tblout))
```



